### ISSUE TITLE
IOError('data byte must be in range 0..127') with Lakh MIDI file

### ISSUE THREAD

**Issue by carlthome** [olemb/mido#63]
I can open [this attached MIDI file](https://github.com/olemb/mido/files/526992/0b0812979e33c132046d24575ff810cf.mid.zip) in timidity and Ardour without issue, but mido is unable to parse it. What's special about it?

E.g.

``` python
from mido import MidiFile
MidiFile('0b0812979e33c132046d24575ff810cf.mid')
```

results in 

``` sh
---------------------------------------------------------------------------
OSError                                   Traceback (most recent call last)
<ipython-input-2-51ef249dd91a> in <module>()
----> 1 MidiFile('0b0812979e33c132046d24575ff810cf.mid')

/home/carl/anaconda3/lib/python3.5/site-packages/mido/midifiles.py in __init__(self, filename, file, type, ticks_per_beat, charset)
    229         elif self.filename is not None:
    230             with io.open(filename, 'rb') as file:
--> 231                 self._load(file)
    232 
    233     def add_track(self, name=None):

/home/carl/anaconda3/lib/python3.5/site-packages/mido/midifiles.py in _load(self, file)
    266 
    267             for i in range(number_of_tracks):
--> 268                 self.tracks.append(self._read_track())
    269                 # Todo: used to ignore EOFError. I hope things still work.
    270 

/home/carl/anaconda3/lib/python3.5/site-packages/mido/midifiles.py in _read_track(self)
    351                 message = self._read_sysex()
    352             else:
--> 353                 message = self._read_message(status_byte)
    354 
    355             message.time = delta

/home/carl/anaconda3/lib/python3.5/site-packages/mido/midifiles.py in _read_message(self, status_byte)
    286         for byte in data_bytes:
    287             if byte > 127:
--> 288                 raise IOError('data byte must be in range 0..127')
    289         return build_message(spec, [status_byte] + data_bytes)
    290 

OSError: data byte must be in range 0..127
```

``` sh
pip show mido

---
Metadata-Version: 2.0
Name: mido
Version: 1.1.17
Summary: MIDI Objects for Python
Home-page: https://mido.readthedocs.io/
Author: Ole Martin Bjorndalen
Author-email: ombdalen@gmail.com
Installer: pip
License: MIT
Location: /home/carl/anaconda3/lib/python3.5/site-packages
Requires: 
Classifiers:
  Development Status :: 5 - Production/Stable
  Intended Audience :: Developers
  Natural Language :: English
  License :: OSI Approved :: MIT License
  Programming Language :: Python
  Programming Language :: Python :: 2.7
  Programming Language :: Python :: 3.2
```

**Comment by carlthome** [comment]
Ping @craffel, same issue for you?

**Comment by craffel** [comment]
This is expected behavior., see https://github.com/craffel/pretty-midi/issues/34  `mido` is strict about data bytes being in 0...127.  `python-midi`, which is what was used during the creation of the Lakh MIDI dataset, just silently passes them (this is bad).  I bet timidity++ and Ardour do something like clip the value to 127 when it's above 127, or something.

**Comment by carlthome** [comment]
Aha, thank you for the clarification! Will Lakh be updated to reflect this? I've found some more mid files with the same problem. Or @olemb, could mido have a quiet mode where it clips instead of raising the IOError?

**Comment by craffel** [comment]
> Will Lakh be updated to reflect this?

It will be whenever I create v0.2, but I don't have a strong reason to do that at this time.  I would guess on the order of a year from now.

> Or @olemb, could mido have a quiet mode where it clips instead of raising the IOError?

Keep in mind that data bytes > 127 are indicators of a corrupt MIDI file.  That is, bug-free MIDI transcription software/hardware should never produce a MIDI data byte larger than 127, so something must be awry.  By default I think raising an Exception is correct; I do think having a mode where it just raised a warning and clipped to 127 could be useful.

**Comment by olemb** [comment]
I agree with @craffel. An exception should be the default behavior.

To see exactly which message is causing trouble you can turn on the (undocumented) debug mode:

```
mido.midifiles.debug()
```

This will print out every byte as it's read from the file along with some other info. Here are the last few bytes before it crashed:

```
  0018c5: 00 
  0018c6: c9 
  0018c7: ff 
```

This is a program change with program 255, which is clearly invalid. I am curious how this happened. Could be a bug in the program that created the file. The vale of 255 is a bit suspicious though.

I'm open to adding a quiet mode but some though needs to be put into how it behaves. In this case clipping to 127 would allow Mido to read the file but it would still select the wrong program.

Most programs seem to be rather forgiving about errors in MIDI files. I would like to investigate how they handle cases such as this one. I am planning to add some test files to Mido. These would be useful here as well.

**Comment by olemb** [comment]
To get full debugging output you need to use `mido.midifiles.MidiFile`. (`debug()` monkey patches `mido.midifiles` but doesn't replace the class anywhere else.)

```
import mido

mido.midifiles.debug()
mid = mido.midifiles.MidiFile('test.mid')
```

**Comment by olemb** [comment]
I've replaced the monkey patch hack with a proper debugging option. You can now do:

```
import mido

mid = mido.MidiFile('test.mid', debug=True)
```

This adds a bit of overhead (due to a few `if self.debug:`) but I think it's worth it. I've also improved the output format a lot.

To get back to the issue, I've tested how different programs handle an out of range data byte. There are four strategies:

* accept the value (Python midi package, Rosegarden)
* clip to 127 (Audacious, aplaymidi)
* skip without warning
* skip the message, but warn on stderr (Ardour)

I think the best option for Mido is to clip to 127. It would be nice to also get a warning but I'm not sure where that warning would go. (Certainly not to stderr.)

As for a name for this option I'm currently considering `ignore_errors`.

@carlthome, thanks for reporting this!

**Comment by craffel** [comment]
> I think the best option for Mido is to clip to 127. It would be nice to also get a warning but I'm not sure where that warning would go. (Certainly not to stderr.)

The Pythonic thing would just be to call `warnings.warn('message')` whenever an invalid data byte was found.

**Comment by carlthome** [comment]
:+1: for warnings.warn instead of stderr (because of builtin suppression support and stuff, e.g. `warnings.filterwarnings("ignore")`), but preferably just print one combined warning after having read a MIDI file, in case there were multiple invalid bytes. Please don't flood the terminal like NumPy's deprecation warnings and so on. :smile:

**Comment by olemb** [comment]
Good point. I'll go for the combined warnings.

There should probably be one warning per type of error, so if you have invalid data bytes and unexpected EOF you will get two warnings.

**Comment by craffel** [comment]
>  but preferably just print one combined warning after having read a MIDI file, in case there were multiple invalid bytes.

This should happen by default.  From the `warnings` docs:
"The disposition of warnings can vary based on the warning category (see below), the text of the warning message, and the source location where it is issued. Repetitions of a particular warning for the same source location are typically suppressed."

**Comment by carlthome** [comment]
Hey, cool! :tada: 

But what's a "particular warning"? Does the message of warnings caused by the same LOC have to be exactly the same for repetitions to be automatically suppressed? E.g. let's say each particular MIDI byte is included in its own warning, will that make `warnings` not suppress them because they are too different?

**Comment by craffel** [comment]
> Does the message of warnings caused by the same LOC have to be exactly the same for repetitions to be automatically suppressed?

Yes, I believe so.

```Python
>>> import warnings
>>> for n in range(10):
...   warnings.warn(n)
...
__main__:2: UserWarning: 0
__main__:2: UserWarning: 1
__main__:2: UserWarning: 2
__main__:2: UserWarning: 3
__main__:2: UserWarning: 4
__main__:2: UserWarning: 5
__main__:2: UserWarning: 6
__main__:2: UserWarning: 7
__main__:2: UserWarning: 8
__main__:2: UserWarning: 9
>>> for n in range(10):
...   warnings.warn('hey')
...
__main__:2: UserWarning: hey
```

**Comment by olemb** [comment]
Cool.

I guess that means we shouldn't include the file position in the message.
:smile:

**Comment by olemb** [comment]
Ah, thanks. Didn't remember that function. I think that's just what we need
here.

Should this mode also handle unexpected end of file while reading a track?
This would allow you to keep the messages that were read up until the end
of the file.

**Comment by craffel** [comment]
> Should this mode also handle unexpected end of file while reading a track?
This would allow you to keep the messages that were read up until the end
of the file.

It seems reasonable to me that there be a catch-all mode that basically just tries really hard to parse the MIDI file despite any invalid things (data bytes larger than 127, unexpected end of file, etc).

**Comment by olemb** [comment]
Sounds good. I haven't come up with a good name for the option. Do you have any suggestions?

There are a few modes that could be useful:

* fail (as today, should be the default)
* warnings.warn()
* ignore errors entirely
* log errors somewhere (file position and error message) so you can inspect the file

I guess the last one is least useful but something to consider still perhaps.

A `warn=False` could be useful even though you can disable warnings in Python.

**Comment by carlthome** [comment]
Many programs call such modes "-q, --quiet" to ignore warnings and "-f, --force" to keep going.

I'd go for always halt on parse error unless -f, and always output warnings unless -q.

For logging, [logging](https://docs.python.org/3/library/logging.html) could be used, but otherwise something like (pseudo-code, beware):
```python
if quiet_mode:
    warnings.filterwarnings('ignore', MidiParseWarning)

if not force:
    warnings.simplefilter('error', MidiParseWarning)

for midi_byte in midi_bytes:
    try:
        read_byte(...)
    except MidiParseWarning as e:
        warnings.warn(e)
```

**Comment by olemb** [comment]
How about this?

```
mid = MidiFile('test.mid', errors='ignore')
```

That would make it easy to add more modes like 'warn', 'log' etc.

Raising a custom exception like `MidiParserWarning` is a good idea. Today it can fail with all kinds of exceptions making them hard to catch. Maybe that should be the default mode?

**Comment by carlthome** [comment]
Actually should have called it `MidiParseError` because an exception shouldn't be conflated with a user preventable warning. Sorry.

Anyway, that parameter looks sweet. I'd use it!

**Comment by olemb** [comment]
Ah, good point. How about `MidiFileError`? I find that parsing makes med think of `mido.Parser`.

Cool. We'll go with the `error='something'` option.

Let's see:

* `errors='raise'`  -> raise `MidiFileError` (default), not sure what to call this one
* `errors='warn'`  -> call `warnings.warn()`
* `errors='list'` -> store errors in a list inside the `MidiFile` object?
* `errors='log'` -> log to `logging` (I don't think we need this if we have `errors='list'`)

**Comment by carlthome** [comment]
Looks good! Though both `errors='list'` and `errors='log'` are [easily covered by `warnings`](https://docs.python.org/3/library/logging.html#logging.captureWarnings) and errors='warn' so I'd absolutely recommend to skip adding that complexity to Mido. At the most, maybe add a little example to the docstring for MidiFile if logging the errors is important.

I'd go for:
 - `errors='raise'` (seems like a good name to me)
 - `errors='warn'`
 - `errors='ignore'`

**Comment by olemb** [comment]
Sounds good. I like avoiding complexity. It will still be a bit tricky to work this in since everything is broken into functions.

If you pass `debug=True` it could print warnings in your debug output, for example:

```
0000f0: c0
0000f1: 23
0000f2: ff
### Warning: clipping out of range data byte
...
### Warning: unexpected end of file while reading track
```

I think it makes sense that `MidiFileError` it inherits from `IOError`:

```
class MidiFileError(IOError):
    pass
```

**Comment by craffel** [comment]
> Looks good! Though both errors='list' and errors='log' are easily covered by warnings and errors='warn' so I'd absolutely recommend to skip adding that complexity to Mido. At the most, maybe add a little example to the docstring for MidiFile if logging the errors is important.

+1

> errors='ignore'

-1, I would highly recommend not reinventing the wheel here.  This would add unneeded complexity to the program, and is not consistent with Python conventions.
```
import warnings
warnings.simplefilter('ignore')
```
in user code is much more straightforward, and people should be used to doing this already.

**Comment by olemb** [comment]
Ah, good point. Maybe `MidiFileError`? Parsing makes me think of the parser
(`mido.Parser`).

Cool. We'll go with `error='something'`.

**Comment by olemb** [comment]
@craffel I guess we only need two modes then: raise and warn. Maybe this should be a flag?

```
MidiFile('test.mid', warn_instead_of_raise=True)
```

Not sure what to call the flag. Certainly not `warn_instead_of_raise`. :smile:

**Comment by craffel** [comment]
I think `errors='warn'` or `errors='raise'` is ok.  Or, if you have a better word for "a corrupt thing found in the MIDI file" than errors, you could use that instead.  Up to you!

**Comment by olemb** [comment]
Hehe, no I don't.

**Comment by carlthome** [comment]
I'd like to make another push for having a builtin `errors='ignore'` mode actually because it should be threadsafe which can be tricky to manage from the outside AFAIK with warnings. `warnings.simplefilter` introduces global state, right? For example, I have a test suite where Mido's parsing should sometimes be strict and sometimes be ignored, and it would be less than ideal to flip a global all over (and a cause for bugs and confusion), not to mention multi-threading stuff that actually becomes impossible to do, no?

Currently I'm just catching the exceptions on my own, but that's what I'd like to get away from. :smile: 
![image](https://cloud.githubusercontent.com/assets/1595907/19802564/9fecd546-9d04-11e6-8867-aa89732a05cd.png)

Also @olemb, regarding changing str to bool when there are just two options, that's bad because you couldn't extend it easily in the future without breaking the API or cluttering the signature. Hopefully the Python community will eventually find their way [here](https://docs.python.org/3/library/enum.html). :dancer:

**Comment by olemb** [comment]
@carlthome I agree. `errors=...` would allow for future expansion.

Actually, adding `errors='ignore'` (and even `errors='print'`) would be easy once `errors='raise'/'warn'` is in place. The tricky bit is that `read_track()`, `read_message()`, `read_sysex()` and `read_meta_message()` have to know about all of this. So far it looks like we'll have to pass it around, for example `read_track(infile, errors='warn')`.

**Comment by olemb** [comment]
I've made an attempt at an implementation.

**Comment by craffel** [comment]
So, I'm happy to defer to @olemb for their preference here, but just responding -

> I'd like to make another push for having a builtin errors='ignore' mode actually because it should be threadsafe which can be tricky to manage from the outside AFAIK with warnings. warnings.simplefilter introduces non thread safe global state even with its context manager.

Do you mean "global state" as in across Python processes/kernels, or just across threads within the same Python process?  It's certainly not global in the sense of across Python processes, but it should apply globally to threads launched within the same process.  I think the use-case of wanting to launch multiple threads simultaneously, where some of them ignore warnings and others produce them, is very far to the edge.  All that aside, I also don't think that a program producing a couple of warnings (keep in mind numpy is an outlier here since they force warnings to appear multiple times) is really a problem in any way.  Also, given that warnings do not change actual program behavior in any way, I don't think the concept of "thread safe" really applies.

> For example, I have a test suite where Mido's parsing should sometimes be strict and sometimes be ignored, and it would be less than ideal to flip a global all over (and a cause for bugs and confusion), not to mention multi-threading stuff that actually becomes impossible to do, no?

I don't see it as better to pass a different argument each time you call the function vs. using the warnings simplefilter.  E.g.
```
warnings.simplefilter('ignore')
function()
function()
warnings.simplefilter('always')
function()
```
vs.
```
function(errors='ignore')
function(errors='ignore')
function(errors='warn')
```
I don't see the latter being any better than the former.  Furthermore, Python has provided us with the incredibly useful `warnings` module and functionality, which is widely used; if Python authors all had to implement their own "do I throw an exception or warn or do nothing" behavior, we'd all end up reinventing the wheel, not to mention with more messy code.  As a Python user I'm used to using `warnings.simplefilter` (or other functionality); I'm not used to passing keyword arguments in to determine warning behavior.

**Comment by olemb** [comment]
I definitely prefer `function()` over `function(errors='ignore'). Is there any way though to get it to raise an exception if `error='raise'` without passing something around?

**Comment by craffel** [comment]
> Is there any way though to get it to raise an exception if error='raise' without passing something around?

Do you mean if you pass the main function a value for `errors` you want functions it calls to get that behavior too?  I think you do need to pass it around.  If you make a PR for your preliminary implementation I can take a look at it to see if there's a way to simplify it.

**Comment by olemb** [comment]
I pushed it to my develop branch. Didn't think of making a pull request.

The funtions need to know about the error mode since byte tests happen
there.

I don't like my new implementation with all the passing around.

Den fre. 28. okt. 2016, 19:38 skrev Colin Raffel <notifications@github.com>:

**Comment by olemb** [comment]
The simplest solution is to just use `warnings.warn()` all of the functions and have no `errors` parameter, although this would not allow you to catch corrupt files in the code.

**Comment by olemb** [comment]
I've reverted my changes and created a pull request as you suggested.

**Comment by carlthome** [comment]
I know Python is mostly used for small projects but the latter is actually __a lot better__ than the former, especially for projects with more than one developer. Refer to [this](https://youtu.be/-FRm3VPhseI?list=PLRjQm0gs0mi9_R450ya905_uU-rHydv1c). 

It's also not a rare use case, for example I'm doing a MIR task currently in which MIDIs are edited, synthesized and put in a queue for training models. The Lakh MIDIs are initially read with suppressed warnings and then edited to simpler MIDI files (removing some octaves, modulation, velocity levels, etc.) and saved before synthesis. Obviously the synthesis code should complain upon finding errors, in case something stupid happened when the original file was edited. 

But regardless, point taken, staying with language conventions is important! I agree that Mido should leave it to the user to handle `warnings`.

@olemb, speaking of dependency injection and passing around the errors parameter (IMO actually a good thing), if you want to avoid it just set `warnings.simplefilter('error', MidiParseWarning)` in __init__.py and make sure all MIDI byte related errors use that particular warning. Then they will be raised as errors and corrupt files can be caught in the code.

Then users can use `warnings.filterwarnings('ignore', MidiParseWarning)` to suppress errors or `warnings.filterwarnings('default', MidiParseWarning)` to just warn. I think this is the most pythonic way. @craffel, agree?

**Comment by craffel** [comment]
> Then users can use warnings.filterwarnings('ignore', MidiParseWarning) to suppress errors or warnings.filterwarnings('default', MidiParseWarning) to just warn. I think this is the most pythonic way. @craffel, agree?

This seems clean to me.

**Comment by carlthome** [comment]
@olemb, just curious on status? Want help?

**Comment by olemb** [comment]
Sorry I haven't gotten back to you.

So the idea is to ditch the `errors=...` and just use `warnings.warn()`? This is certainly a lot cleaner and simpler.

A couple of thoughs:

* The default will now be to clip the bytes to 0-127 and warn the user. This means files with recoverable errors will load without any extra work. Sounds like a good default.

* When using `debug=True` it would be good to have the warnings as part of the output so you can see where they are. It looks like this can be done with the `warn` module (by showing all instances of the warning and redirecting to stdout). I'm not sure it would be a good idea to turn this on then `debug=True` is passed as it messes with the user's settings.

**Comment by olemb** [comment]
I'm wondering if we should just clip the bytes by default and consider adding a warning or some option later. This way the files would load.

@cjappl Any thoughs?

**Comment by cjappl** [comment]
I think they should probably go together, as we wouldn't want to have something silently editing the midi files.

I'd say do whatever is the simplest (warnings.warn I'd guess) and document that:

1) A midi's files notes will be limited to 0-127 if they are outside this range
2) How to silence the warnings in your end program if you want to (using one of the warning filters)

I think if these are put somewhere prominent in the documentation, it'll not take people by surprise when they see these warnings, but the files will still load as intended.

Another option (just to add to the pile here) is to keep the default behavior but add a: "clip=True/False" option. If specified, any value would be limited to 0-127. This would solve the original problem, and it'd have to be an active choice by the programmer, so you couldn't say he wasn't aware there was a chance his midi was edited.

I'd argue the "cleanest" solution is to throw an error when there is a corupt midi file (as the package already does). If there is push for limiting the values to the proper range, just giving the boolean flag may be the clean/easy way without fussing around with warnings and filtering them.

**Comment by carlthome** [comment]
> I'd argue the "cleanest" solution is to throw an error when there is a corupt midi file (as the package already does). If there is push for limiting the values to the proper range, just giving the boolean flag may be the clean/easy way without fussing around with warnings and filtering them.

:+1: 

I'd like to know if my MIDI files are wonky.

**Comment by cjappl** [comment]
Possible solution here in pull request #110

**Comment by cjappl** [comment]
This option should be documented somewhere but wanted to check in before just throwing it in somewhere. It seems like on the library reference page [here](http://mido.readthedocs.io/en/latest/lib.html#mido.MidiFile) there is currently no documentation of the parameters given in the initializer. Adding these may fall under a different task

**Comment by olemb** [comment]
Good plan. Let's do it this way.

You're right. The initializer needs documentation. I guess it may be missing for other classes as well. It's definitely a new issue though.

Btw, do you know a good way to document properties? I tried to add doc strings to the getters but they never showed up.

**Comment by cjappl** [comment]
I don't off the top of my head. I know there is a way to document decorated functions using functools.wrap.

[Heres](https://stackoverflow.com/a/17522081/7299836) one way apparently. Haven't tried it myself though

**Comment by cjappl** [comment]
I believe we can close this with PR #110 being merged. I'll make an issue for documenting the changes

**Comment by olemb** [comment]
We need to document the `clip` option.

**Comment by cjappl** [comment]
Agreed, should it be a separate issue or rolled into the over-arching issue of #111?

Also in terms of housekeeping stuff, I wanted to do a PR where I moved the tests out into a "test" directory. Do you have any strong feelings about that? I personally like to keep the tests separate from the source.

**Comment by olemb** [comment]
I think separate issue.

I'm open to moving the tests to a "test" directory but we need to discuss the structure a bit. Can you make a new issue? ("Reorganizing Tests" or something.)

**Comment by cjappl** [comment]
Alright cool. Let me file some issues really quick then we can maybe close this one?

**Comment by olemb** [comment]
Great! Closing it.

**Comment by diegodlh** [comment]
I wonder if these lines in midifiles/midifiles.py, which check if data_bytes are greater than 127 if clip=off:
https://github.com/mido/mido/blob/89ae5a4ddd6f51e8b0911b5ca343cb2016030446/mido/midifiles/midifiles.py#L133-L136
Aren't redundant with these other lines in messages/checks.py, which seem to do the same:
https://github.com/mido/mido/blob/89ae5a4ddd6f51e8b0911b5ca343cb2016030446/mido/messages/checks.py#L56-L57

In fact, if I comment these lines in midifiles/midifiles.py, files with data_bytes > 127 end up raising the ValueError specified in messages/checks.py.
